currencies/rate_converter_test.go

```diff
@@ -0,0 +1,477 @@
+	}
+
+	// Execute:
+	rateConverter := currencies.NewRateConverter(
+		mockedHttpClient,
+		requestURL,
+		time.Duration(0)*time.Millisecond,
+		time.Duration(5)*time.Minute,
+	)
+
+	// Verify:
+	// Check for the next 1 second that no fetch was triggered
+	time.Sleep(1 * time.Second)
+
+	assert.False(t, mockedHttpClient.VerifyUrlCalled(requestURL), "sync URL `%s` was called", requestURL)
+	assert.Equal(t, (time.Time{}), rateConverter.LastUpdated(), "LastUpdated() shouldn't be set")
+	assert.Nil(t, rateConverter.Rates(), "Rates should be nil")
+}
+
+func TestRates(t *testing.T) {
+
+	// Setup:
+	testCases := []struct {
+		from         string
+		to           string
+		expectedRate float64
+		hasError     bool
+	}{
+		{from: "USD", to: "GBP", expectedRate: 0.77208, hasError: false},
+		{from: "GBP", to: "USD", expectedRate: 1.2952, hasError: false},
+		{from: "GBP", to: "EUR", expectedRate: 0, hasError: true},
+		{from: "CNY", to: "EUR", expectedRate: 0, hasError: true},
+		{from: "", to: "EUR", expectedRate: 0, hasError: true},
+		{from: "CNY", to: "", expectedRate: 0, hasError: true},
+		{from: "", to: "", expectedRate: 0, hasError: true},
+	}
+
+	requestURL := "currency.fake.com/latest.json"
+	responseBody := `{
+		"dataAsOf":"2018-09-12",
+		"conversions":{
+			"USD":{
+				"GBP":0.77208
+			},
+			"GBP":{
+				"USD":1.2952
+			}
+		}
+	}`
+
+	mockedHttpClient := &HttpClientMock{
+		responseStatusCode: http.StatusOK,
+		responseBody:       responseBody,
+		err:                nil,
+	}
+
+	// Execute:
+	ticks := make(chan int)
+	rateConverter := currencies.NewRateConverterWithNotifier(
+		mockedHttpClient,
+		requestURL,
+		time.Duration(100)*time.Millisecond,
+		time.Duration(5)*time.Minute,
+		ticks,
+	)
+	rates := rateConverter.Rates()
+
+	// Let the currency converter ticks 1 time before to stop it
+	select {
+	case <-ticks:
+		rateConverter.StopPeriodicFetching()
+	}
+
+	// Verify:
+	assert.NotNil(t, rates, "rates shouldn't be nil")
+	for _, tc := range testCases {
+		rate, err := rates.GetRate(tc.from, tc.to)
+
+		if tc.hasError {
+			assert.NotNil(t, err, "err shouldn't be nil")
+			assert.Equal(t, float64(0), rate, "rate should be 0")
+		} else {
+			assert.Nil(t, err, "err should be nil")
+			assert.Equal(t, tc.expectedRate, rate, "rate doesn't match the expected one")
+		}
+	}
+}
+
+func TestRates_EmptyRates(t *testing.T) {
+
+	// Setup:
+	requestURL := "currency.fake.com/latest.json"
+
+	mockedHttpClient := &HttpClientMock{
+		responseStatusCode: http.StatusOK,
+		responseBody:       "",
+		err:                nil,
+	}
+
+	// Execute:
+	// Will try to fetch directly on method call but will fail
+	rateConverter := currencies.NewRateConverter(
+		mockedHttpClient,
+		requestURL,
+		time.Duration(100)*time.Millisecond,
+		time.Duration(5)*time.Minute,
+	)
+	defer rateConverter.StopPeriodicFetching()
+	rates := rateConverter.Rates()
+
+	// Verify:
+	assert.Nil(t, rates, "rates should be nil")
+}
+
+type HttpClientMock struct {
```

Comment:
Any reason to not use https://golang.org/pkg/net/http/httptest/ ?

Reply:
No, not really, I will replace this mock with the httptest package